tls/tls_srv.c

```diff
@@ -0,0 +1,2331 @@
+
+	return 0;
+}
+
+static int
+ttls_write_server_hello_done(TlsCtx *tls, struct sg_table *sgt,
+			     unsigned char **in_buf)
+{
+	TlsIOCtx *io = &tls->io_out;
+	unsigned char *p = *in_buf;
+
+	T_DBG("sending ServerHelloDone\n");
+
+	io->msglen = TTLS_HS_HDR_LEN;
+	ttls_write_hshdr(TTLS_HS_SERVER_HELLO_DONE, p + TLS_HEADER_SIZE,
+			 TTLS_HS_HDR_LEN);
+
+	*in_buf += TLS_HEADER_SIZE + TTLS_HS_HDR_LEN;
+	sg_set_buf(&sgt->sgl[sgt->nents++], p, *in_buf - p);
+	get_page(virt_to_page(p));
+
+	__ttls_add_record(tls, sgt, sgt->nents - 1, p);
+
+	return 0;
+}
+
+static int
+ttls_parse_client_dh_public(TlsCtx *tls, unsigned char **p,
+			    const unsigned char *end)
+{
+	int r = TTLS_ERR_FEATURE_UNAVAILABLE;
+	size_t n;
+
+	/* Receive G^Y mod P, premaster = (G^Y)^X mod P. */
+	if (*p + 2 > end) {
+		T_DBG("bad client dh key exchange message\n");
+		return TTLS_ERR_BAD_HS_CLIENT_KEY_EXCHANGE;
+	}
+
+	n = ((*p)[0] << 8) | (*p)[1];
+	*p += 2;
+
+	if (*p + n > end) {
+		T_DBG("bad client key exchange message\n");
+		return TTLS_ERR_BAD_HS_CLIENT_KEY_EXCHANGE;
+	}
+
+	if ((r = ttls_dhm_read_public(&tls->hs->dhm_ctx, *p, n))) {
+		T_DBG("cannot read dhm public, %d\n", r);
+		return TTLS_ERR_BAD_HS_CLIENT_KEY_EXCHANGE_RP;
+	}
+
+	*p += n;
+
+	TTLS_DEBUG_MPI("DHM: GY", &tls->hs->dhm_ctx.GY);
+
+	return r;
+}
+
+static int
+ttls_parse_encrypted_pms(TlsCtx *tls, const unsigned char *p,
+			 const unsigned char *end)
+{
+	unsigned char mask;
+	int r;
+	unsigned int diff;
+	size_t len = ttls_pk_get_len(ttls_own_key(tls));
+	size_t i, peer_pmslen;
+	unsigned char *pms = tls->hs->premaster;
+	unsigned char ver[2], fake_pms[48], peer_pms[48];
+
+	BUILD_BUG_ON(sizeof(tls->hs->premaster) < 0
+		     || sizeof(tls->hs->premaster) < 48);
+	if (!ttls_pk_can_do(ttls_own_key(tls), TTLS_PK_RSA)) {
+		T_DBG("got no RSA private key\n");
+		return TTLS_ERR_PRIVATE_KEY_REQUIRED;
+	}
+
+	/* Decrypt the premaster using own private RSA key. */
+	if (*p++ != ((len >> 8) & 0xFF) || *p++ != (len & 0xFF)
+	    || p + len != end)
+	{
+		T_DBG("bad client key exchange message\n");
+		return TTLS_ERR_BAD_HS_CLIENT_KEY_EXCHANGE;
+	}
+
+	ver[0] = TTLS_MAX_MAJOR_VERSION;
+	ver[1] = tls->conf->max_minor_ver;
+
+	/*
+	 * Protection against Bleichenbacher's attack: invalid PKCS#1 v1.5
+	 * padding must not cause the connection to end immediately; instead,
+	 * send a bad_record_mac later in the handshake.
+	 * Also, avoid data-dependant branches here to protect against
+	 * timing-based variants.
+	 */
+	ttls_rnd(fake_pms, sizeof(fake_pms));
+
+	r = ttls_pk_decrypt(ttls_own_key(tls), p, len, peer_pms, &peer_pmslen,
+			    sizeof(peer_pms));
+
+	diff  = (unsigned int) r;
+	diff |= peer_pmslen ^ 48;
+	diff |= peer_pms[0] ^ ver[0];
+	diff |= peer_pms[1] ^ ver[1];
+	T_DBG("client key exchange message diff=%x\n", diff);
+
+	tls->hs->pmslen = 48;
+	/* mask = diff ? 0xff : 0x00 using bit operations to avoid branches */
+	mask = - ((diff | - diff) >> (sizeof(unsigned int) * 8 - 1));
```

Comment:
`diff | -diff` seems to be invalid operation here. This operation always result as `0xffff_ffff `.